posts/inside-rust/2020-02-07-intro-rustc-self-profile.md

```diff
@@ -0,0 +1,269 @@
++-----------------------------------------------+-----------+-----------------+----------+------------+------------+--------------+-----------------------+
+...
+Total cpu time: 11.440758871s
+```
+
+The output is sorted by the self time (time spent in the query or activity but not other queries or activites called by itself).
+As you can see, most of the compilation time is spent in LLVM generating the binary code for the executable.
+
+## `flamegraph`
+
+As you may have guessed, `flamegraph` will produce a [flame graph] of the profiling data.
+To run the tool, we'll pass just the filename without a file extension like we did for `summarize`:
+
+```sh
+$ ../measureme/target/release/flamegraph regex-17088
+```
+
+This will create a file called `rustc.svg` in the working directory:
+
+![Image of flamegraph output][flame graph img]
+
+[Click here] to try the interactive svg.
+
+## `crox`
+
+This tool processes self-profiling data into the JSON format that the Chromium profiler understands.
+You can use it to create a graphical timeline showing exactly when various traced events occurred.
+
+In this section, we'll cover a few different modes `crox` can run in such as profiling an entire crate compilation including dependencies and filtering out small events.
+Let's get started with the basics!
+
+### Basic usage
+
+To run the tool, we'll just pass the filename without a file extension like we've done before:
+
+```sh
+$ ../measureme/target/release/crox regex-17088
+```
+
+This creates a file called `chrome_profiler.json` in the working directory.
+To open it, we'll use the regular Chromium performance tools you might already be familar with:
+
+1. Open Chrome
+2. Open the Developer Tools console by pressing `Ctrl` + `Shift` + `i` (Windows/Linux) or `Cmd` + `Option` + `i` (macOS)
+3. Click the Performance tab at the top of the console.
+4. Click the "Load profile" button which looks like an arrow pointing up.
+5. Select the `chrome_profiler.json` file we crated.
+
+You should now see something similar to this:
+
+![Image of chrome profiler][chrome profiler img1]
+
+You can use the scroll wheel on a mouse or the appropriate gesture on a touchpad to zoom in or out of the timeline.
+
+### Filtering short events
+
+If the `chrome_profiler.json` file gets too large, the normal Chromium performance tools have issues opening the file.
+One easy way to deal with this is to tell `crox` to remove events shorter than a chosen duration:
+
+```sh
+$ ../measureme/target/release/crox --minumum-duration 2 regex-17088
+```
+
+Filtering out events less than 2 microseconds shrinks our `chrome_profiler.js` file from 27mb to 11mb.
+
+### Capturing event arguments
+
+The self-profiler can be configured to record event arguments during compilation.
+For example, queries will include their query key.
+This fuctionality is turned off by default because it significantly increases the self-profiler overhead.
```

Comment:
I'm wondering what the overhead is. File sizes are 2-3x, but runtime is not affected as much, I think.

Reply:
Looks like the overhead is ~5% on most of the crates I've looked at. I'll change this to omit the "significantly" adjective.